File: midi.rb
module Midi
    def start?
        self[0].ord == 0xfa
    end
    def stop?
        self[0].ord == 0xfc
    end	
    def pulse?
        self[0].ord == 0xf8
    end
    def clock?
        [0xfa,0xfc,0xf8].include?(self[0].ord)
    end
    def note?
        [0x90,0x80].include?(self[0].ord & 0xF0)
    end
    
    def note_on?
        self[0].ord & 0xF0 == 0x90 
    end
    
    def note_off?
        self[0].ord & 0xF0 == 0x80 
    end
    
    def chan
        return (self[0].ord & 0xF) + 1 if note?
        raise "not a channel message"
    end

    def pitch 
        return self[1].ord if note?
        raise "not a note, pitch not applicable"
    end
    
    def velo
        return self[2].ord if note?
        raise "not a note, velocity not applicable"
    end
    
    def pp
        if note_on?
            printf("note on  ch:%-2d %-3d %-3d", chan, pitch, velo)
        elsif note_off?
            printf("note off ch:%-2d %-3d %-3d", chan, pitch, velo)
        elsif start?
            printf("start")
        elsif stop?
            printf("stop")
        elsif click?
            printf("click")
        else
            self.bytes.each { |c| printf "0x%x ", c.ord }
        end
        puts
        return self
    end
end

File: midi.js
/*
 * midi Input and Output ports
 */

const midi   = require('midi');
const Events = require('events');
const _      = require('lodash');

const selectPort = (ports, name) => {
    let list = listPorts(ports);
    if (name in list) {
        ports.openPort(list[name]);
    } else {
        console.log(list);
        throw new Error("Unknown device " + name);
    }
}

const listPorts = (ports) => {
    let list = {};
    for (let i = 0; i < ports.getPortCount(); i++) {
        list[ports.getPortName(i)] = i;
    }
    return list;
}

class Input extends Events {
    constructor(name) {
        super();
        const input = new midi.input();
       
        input.on('message', (time, message) => {
            /*
             * 0xFA, 250, 0b11111010 - transport start
             * 0xFC, 252, 0b11111100 - transport stop
             * 0xF8, 248, 0b11111000 - clock pulse, 24 pulses per quarter note
             */
            if (message[0] === 0xFA) { this.emit("start"); }
            if (message[0] === 0xFC) { this.emit("stop");  }
            if (message[0] === 0xF8) { this.emit("clock"); }
        });

        selectPort(input, name)
        input.ignoreTypes(true, false, true);
    }
}

class Output {
    constructor(name) {
        this.queue = [];
        this.output = new midi.output();
        selectPort(this.output, name);
    }

    step() {
        let message = this.queue.shift();
        if(message !== undefined) {
            do {
                this.output.sendMessage(message.splice(0,3));
            } while (message.length > 0);
        }
    }

    play({ note, channel = 1, velocity = 90, gate = 0}) {
        let arr = [[0x90 + channel - 1,note,velocity]]
        if (gate !== 0) {
            arr.push(...Array(gate - 1),[0x80 + channel - 1,note,0]);
        }
        this.queue = _.zipWith(this.queue, arr, (a, b) => {
            if (a == undefined) { return b; }
            if (b == undefined) { return a; }
            return [...a, ...b];
        });
    }
}

module.exports = { Input, Output };
